version: '3.8'

services:
  frontend:
    build:
      context: ./src/frontend
      dockerfile: Dockerfile
      args:
        VITE_OPENCAGE_TOKEN: ${VITE_OPENCAGE_TOKEN}
    working_dir: /app
    command: ["npm", "run", "preview", "--", "--port", "5173", "--host"]
    ports:
      - "5173:5173"
    environment:
      - API_HOST=host.docker.internal
    networks:
      - ml-network

  nginx:
    image: nginx:alpine
    depends_on:
      - api
      - frontend
    ports:
      - "3000:80"
    volumes:
      - ./src/proxy/nginx.conf:/etc/nginx/conf.d/default.conf
    extra_hosts:
      - "host.docker.internal:host-gateway"  # Linux compatibility
    networks:
      - ml-network

  mlflow:
    image: python:3.9-slim
    container_name: mlflow-server
    ports:
      - "5000:5000"
    environment:
      - MLFLOW_TRACKING_URI=http://localhost:5000
    command: >
      bash -c "pip install mlflow psycopg2-binary boto3 && mlflow server --host 0.0.0.0 --port 5000 --backend-store-uri sqlite:///mlflow.db"
    networks:
      - ml-network

  api:
    build:
      context: .
      dockerfile: Dockerfile.api
    container_name: fraud-detection-api
    restart: on-failure
    ports:
      - "5001:5001"
    volumes:
      - ./:/app
    environment:
      - ML_FLOW_URI=http://mlflow:5000
    depends_on:
      - mlflow
    networks:
      - ml-network

  minio:
    image: minio/minio
    container_name: minio
    ports:
      - "9000:9000"   # S3 API
      - "9001:9001"   # Web Console
    environment:
      MINIO_ROOT_USER: admin
      MINIO_ROOT_PASSWORD: admin123
    volumes:
      - minio-data:/data
    command: server /data --console-address ":9001"
    restart: unless-stopped
  minio-init:
    image: minio/mc
    depends_on:
      - minio
    entrypoint: >
      /bin/sh -c "
      sleep 5 &&
      mc alias set local http://minio:9000 admin admin123 &&
      mc mb -p local/cc-fraud-detection-2025 ||
      echo 'Bucket already exists'
      "

networks:
  ml-network:
  
volumes:
  frontend-dist:
  minio-data: